Add prometheus url to ui config

jbertran opened the pull request:
**Component**:

salt, ui, monitoring
<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: We want to supply the UI ConfigMap with the Prometheus URL, in order for it to use Prometheus for monitoring purposes.

**Summary**:

* Make endpoint lookups namespace-sensitive in the ext_pillar
* Add the prometheus_url key to the config map
* Add a NodePort service to expose Prometheus on the node

**Acceptance criteria**: After a successful deployment, we have the `prometheus_url` key in `<UI_URL>/config.json`, and the value of this key is an accessible URL to Prometheus on the node where it runs.

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #1131 
<!-- If you want to refer to an issue while not closing it, use:

-->

Commit:
pillar: make endpoint lookup namespaced, add prometheus endpoint

Issue: #1131

Commit:
ui deployment: add prometheus endpoint to ConfigMap

Issue: #1131

bert-e commented:
# Hello jbertran,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/1131-add-prometheus-url-to-ui-config` and the
destination branch `development/2.0`.

Please resolve the conflict on **the feature branch** (`improvement/1131-add-prometheus-url-to-ui-config`).

```bash
 $ git fetch
 $ git checkout origin/improvement/1131-add-prometheus-url-to-ui-config
 $ git merge origin/development/2.0
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/1131-add-prometheus-url-to-ui-config
```

Commit:
monitoring: add NodePort exposing Prometheus to its host node

In the absence of a load balancer, we hackishly use a NodePort
to expose Prometheus on the host network.

Issue: #1131

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/1131-add-prometheus-url-to-ui-config` into
`development/2.0`
* [w/2.1/improvement/1131-add-prometheus-url-to-ui-config](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/1131-add-prometheus-url-to-ui-config) will be merged into `development/2.1` 
* [w/2.2/improvement/1131-add-prometheus-url-to-ui-config](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/1131-add-prometheus-url-to-ui-config) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @jbertran

* @monpote

* @alxf

* @gdemonet

* @lucieleonard

* @zempashi

* @carlito-scality

* @alexandre-allard-scality

* @nootal

* @nicolast

* @thomasdanan

* @tcarmet

* @slaperche-scality

* @teddyandrieux

* @sayf-eddine-scality

* @ramanduh

bert-e commented:
# History mismatch



Merge commit #f4673c30d878268bb8360f00765e43997bc7c53f on the integration branch
`w/2.1/improvement/1131-add-prometheus-url-to-ui-config` is merging a branch which is neither the current
branch `improvement/1131-add-prometheus-url-to-ui-config` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/1131-add-prometheus-url-to-ui-config` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #a80c5454bb8fa27f6b415d11ebf6124166d7b457 on the integration branch
`w/2.1/improvement/1131-add-prometheus-url-to-ui-config` is merging a branch which is neither the current
branch `improvement/1131-add-prometheus-url-to-ui-config` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/1131-add-prometheus-url-to-ui-config` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Commit:
monitoring: fix stray case mismatch in Grafana manifest

Issue: #1131

Commit:
pillar: add nodeport host IP endpoints

slaperche-scality commented on salt/_pillar/metalk8s_endpoints.py:
```diff
@@ -16,11 +16,11 @@ def __virtual__():
 
 
-def service_endpoints(service, kubeconfig):
+def namespaced_service_endpoints(service, namespace, kubeconfig):
```
minor: You could have kept the old name

slaperche-scality commented on salt/_pillar/metalk8s_endpoints.py:
```diff
@@ -16,6 +16,48 @@ def __virtual__():
         return __virtualname__
 
 
+def namespaced_service_endpoints_nodeport(service, namespace, kubeconfig):
```
ditto, I'm not sure that it's useful to add the `namespaced` prefix: eveything seems to be namespaced in k8s anyway.

slaperche-scality commented on salt/_pillar/metalk8s_endpoints.py:
```diff
@@ -16,6 +16,48 @@ def __virtual__():
+    node_name = endpoint.get('node_name')
+
+    if not node_name:
+        log.exception('Cannot get `node_name` from %s.', endpoint)
```
why `log.exception`, there is no exception raised here?

gdemonet replied:
Yes, `log.error` would be better. Even then, we should start implementing the `_errors` key as described in #1018

slaperche-scality commented on salt/metalk8s/addons/ui/deployed.sls:
```diff
@@ -3,7 +3,7 @@
 {%- set apiserver = 'https://' ~ pillar.metalk8s.api_server.host ~ ':6443' %}
 {%- set saltapi = 'http://' ~ pillar.metalk8s.endpoints['salt-master'].ip ~ ':' ~ pillar.metalk8s.endpoints['salt-master'].ports.api %}
-{%- set prometheus = 'http://' ~ pillar.metalk8s.endpoints['prometheus'].ip ~ ':' ~ pillar.metalk8s.endpoints['prometheus'].ports.api %}
+{%- set prometheus = 'http://' ~ pillar.metalk8s.endpoints['prometheus'].ip ~ ':' ~ '30222' %}
```
Why this change?

TeddyAndrieux commented on salt/metalk8s/addons/monitoring/prometheus/exposed.sls:
```diff
@@ -0,0 +1,28 @@
+        type: NodePort
+  require:
+    - pkg: Install Python Kubernetes client
+    - salt: metalk8s.addons.monitoring.prometheus.deployed
```
These requires are useless here as it's a yaml key in `Expose Prometheus` and btw you should not require something not in the current sls (or in an included one)

TeddyAndrieux commented on buildchain/buildchain/salt_tree.py:
```diff
@@ -150,6 +150,7 @@ def _run(self) -> None:
     Path('salt/metalk8s/addons/monitoring/kube-state-metrics/deployed.sls'),
     Path('salt/metalk8s/addons/monitoring/node-exporter/deployed.sls'),
     Path('salt/metalk8s/addons/monitoring/prometheus/deployed.sls'),
+    Path('salt/metalk8s/addons/monitoring/prometheus/exposed.sls'),
```
Why separate `deployed` and `exposed` ? Currently the `exposed` state "deploy" a config map :stuck_out_tongue: 
We said few week ago that all state that should only be launched on the master should be called `deployed.sls`

gdemonet replied:
Agreed, though `deployed` is actually the upstream YAML that gets rendered through Salt into proper states... Maybe the upstream one should get renamed and `include`d in `deployed.sls`

jbertran replied:
The idea here comes from that this state was to a large extent autogenerated from [coreos upstream](https://github.com/coreos/kube-prometheus), so I want to separate them from what's really ours.

More precisely:

1. We're not the authors per se of the contents of `deployed`, whereas we are in `exposed` or however we choose to call the ingress deployment.
2. Being "pulled" from upstream, the `deployed` files should not be modified after their initial version adjustment, whereas the `exposed` files could be, so they're not similar lifecycle-wise.

If we want to keep the naming convention on master-only states though, we can definitely have a `metalk8s.addons.monitoring.prometheus-ingress` or similar state instead - I wasn't aware of it!

jbertran replied:
From discussion with @gdemonet, another sane option would be to store the "upstream" states in `upstream.sls` and just include these states in `deployed.sls`, where we could have our custom states for exposing services to the outside.

TeddyAndrieux commented on salt/metalk8s/deployed.sls:
```diff
@@ -4,10 +4,10 @@ include:
   - metalk8s.kubernetes.coredns.deployed
   - metalk8s.repo.deployed
   - metalk8s.salt.master.deployed
-  - metalk8s.addons.ui.deployed
```
For me you should not remove this from here just put it in the right order here and if needed add a logic in the `prometheus` state to check the status of the ConfigMap before finishing the state.

Maybe just add a state in `prometheus.deployed` that retrieve the configmap from k8s api with a retry just after the configmap deployement

TeddyAndrieux replied:
@jbertran ok you just have to add endpoint like `salt-master` and `repositories` in `pillar_data` in `salt/metalk8s/orchestrate/bootstrap/init.sls`

gdemonet commented on salt/_pillar/metalk8s_endpoints.py:
```diff
@@ -16,11 +16,53 @@ def __virtual__():
 
 
-def service_endpoints(service, kubeconfig):
+def namespaced_service_endpoints_nodeport(service, namespace, kubeconfig):
```
Hum, no let's not do this _workaround_ thing in the ext_pillar. The service IP is still the cluster IP, so we shouldn't get out of our way just to _make things work_. Instead, you should use `grains['workload_plane_ip']` in `addons.ui.deployed`, until we can reach the real service address through some LB/Ingress.

jbertran replied:
From f2f discussions. This is indeed ugly but we're keeping it for the short duration between here and LB/Ingress addition.

Commit:
monitoring: move upstream autogenerated states to `upstream.sls`

Issue: #1131

Commit:
monitoring: include `upstream` states in `deployed` states

Issue: #1131

Commit:
monitoring: add prometheus IP to bootstrap pillar data

Issue: #1131

MonPote commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.1/improvement/1131-add-prometheus-url-to-ui-config` with contents from `improvement/1131-add-prometheus-url-to-ui-config`
and `development/2.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.1/improvement/1131-add-prometheus-url-to-ui-config origin/development/2.1
 $ git merge origin/improvement/1131-add-prometheus-url-to-ui-config
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.1/improvement/1131-add-prometheus-url-to-ui-config
```

jbertran commented:
Done little bot

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.2/improvement/1131-add-prometheus-url-to-ui-config` with contents from `w/2.1/improvement/1131-add-prometheus-url-to-ui-config`
and `development/2.2`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.2/improvement/1131-add-prometheus-url-to-ui-config origin/development/2.2
 $ git merge origin/w/2.1/improvement/1131-add-prometheus-url-to-ui-config
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.2/improvement/1131-add-prometheus-url-to-ui-config
```

MonPote reviewed (dismissed):
LGTM

jbertran commented:
Fixed, bert-e boy

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @alexandre-allard-scality

* @gdemonet

* @zempashi

* @sayf-eddine-scality

* @slaperche-scality

* @alxf

* @monpote

* @tcarmet

* @carlito-scality

* @ramanduh

* @cmonfort

* @thomasdanan

* @lucieleonard

* @nicolast

* @teddyandrieux

* @nootal





The following reviewers are expecting changes from the author, or must review again:

* @teddyandrieux

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @alexandre-allard-scality

* @gdemonet

* @zempashi

* @sayf-eddine-scality

* @slaperche-scality

* @alxf

* @monpote

* @tcarmet

* @carlito-scality

* @ramanduh

* @cmonfort

* @thomasdanan

* @lucieleonard

* @nicolast

* @teddyandrieux

* @nootal

jbertran commented:
@bert-e approve

bert-e commented:
# Conflict with a changeset in the queue



The changeset in this pull request conflicts with another changeset
already in the queue. Please wait for the current queue to merge into
the development branch. The conflict will then appear in this pull
request and can be sorted on the feature branch directly.

This changeset has *not* been added to the queue. 



*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #449dd2d4000439576305f5ab6d369f40bbce4a2f on the integration branch
`w/2.1/improvement/1131-add-prometheus-url-to-ui-config` is merging a branch which is neither the current
branch `improvement/1131-add-prometheus-url-to-ui-config` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/1131-add-prometheus-url-to-ui-config` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

jbertran commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.1/improvement/1131-add-prometheus-url-to-ui-config` with contents from `improvement/1131-add-prometheus-url-to-ui-config`
and `development/2.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.1/improvement/1131-add-prometheus-url-to-ui-config origin/development/2.1
 $ git merge origin/improvement/1131-add-prometheus-url-to-ui-config
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.1/improvement/1131-add-prometheus-url-to-ui-config
```





*The following options are set:* **approve**

jbertran commented:
2.1 done bert-e, gimme 2.2 now

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.2/improvement/1131-add-prometheus-url-to-ui-config` with contents from `w/2.1/improvement/1131-add-prometheus-url-to-ui-config`
and `development/2.2`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.2/improvement/1131-add-prometheus-url-to-ui-config origin/development/2.2
 $ git merge origin/w/2.1/improvement/1131-add-prometheus-url-to-ui-config
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.2/improvement/1131-add-prometheus-url-to-ui-config
```





*The following options are set:* **approve**

jbertran commented:
Gottem berte-e boy. Let's merge!

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/3dd21c6432b65f2efc5eea2d7aafe61a3b0216aa) did not succeed in branch w/2.1/improvement/1131-add-prometheus-url-to-ui-config.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye jbertran.
